index.html

```diff
@@ -1851,39 +1850,321 @@ <h5><code>invokeaction</code></h5>
           </a>).
         </section>
       </section>
+
       <section id="events">
         <h4>Events</h4>
-        <section class="ednote">
+        <p>
+          The Core Profile uses Server-Sent Events [[EVENTSOURCE]] as a
+          mechanism for Consumers to subscribe to events emitted by a Web Thing.
+        </p>
+        <p class="note">
```

Comment:
> In addition to this note, which should reference section 4. Should it also exclude the processing model, which refers to section 4?

As I tried to explain on the Architecture call, this is why I wanted to refer to the Server-Sent Events specification as a whole rather than trying to slice it up, because the way the specification is written tightly couples the scripting API with the wire protocol. E.g. We can't exclude the section 5 (Processing Model) because it's a core part of the protocol definition, but it does make reference to section 4 (the EventSource interface). Sections 1, 7, 8, 9 and 10 also reference section 4.

That's a shame, because there's really nothing JavaScript-specific about the Server-Sent events protocol, demonstrated by the fact that there are existing SSE client implementations in Python, Java, Rust etc. which provide their own language-specific APIs.

The specification does allude to this by mentioning in the Introduction that the client API can emulated using XMLHttpRequest (i.e. a generic HTTP client library), but that in the case of web browsers an implementation of the `EventSource` API can provide additional optimisations.

> The processing model is a bit vague, e.g.
> 
> > For HTTP connections, the Accept header may be included; if included, it must contain only formats of event framing that are supported by the user agent (one of which must be text/event-stream, as described below).
> > Which Accept header are we using?

I don't think this is particularly vague. It's just saying that an `Accept` header in the request used to initiate a connection MUST include `text/event-stream`, but it may also include other formats if the client supports them. It is a bit loose in that it it only says an `Accept` header MAY be included, but what I read from this that if an `Accept` header isn't included then no content negotiation is required and the server should just return `text/event-stream` regardless.

> I believe we we need to create a few normative assertions that clearly states which parts of SSE MUST be implemented and provides a few additional clarifications.

I'm not really sure what to suggest here unless we plan on forking the whole Server-Sent Events specification with a version which is less tightly coupled with the `EventSource` JavaScript API. Do you have a suggestion of alternative text which would make this more clear?

Reply:
A careful reading of the SSE spec unfortunately showed a couple of significant issues/gaps:

Chaper 5 of this specification defines a **normative processing model** on the user agent, which assumes a JavaScript based browser implementation and mandates a specific implementation methodology e.g. describing task queues etc.
 which is appropriate for a browser. 
 However these requirements put unnecessary requirements on non-browser consumers and should be excluded. We need however to incorporate the protocol handling part from that chapter, but these are only a few paragraphs.
 
 Chapter 6 defines the lexical format "text/event-stream". It is strange that the name field names may include non-printable characters as well as space, i.e. a name field may actually be a name that contains spaces.  These cases are however excluded in 
 Chapter 7

Chapter 7 defines 4 field names of the event stream and their interpretation
This is a plain text serialisation of event types and event data where we need to define a mapping of the data schema, which is defined by the event affordance of the TD spec (section 5.3.1.5) to this serialised text stream.
Event types are defined in SSE, however we need to defin a mapping as well.
It would be appropriate to consider 
The retry field leaves some room for interpretation (which unit?), also the permitted values of the id field needs clarification.  

Chapter 9 assumes a push proxy - we need to discuss details further.
 
Chapter 10 is probably not  required by the WoT profile.

With all this taken into account, we should either cherry-pick and provide significant clarifications on the SSE model 
and define a mapping of the event dataschema to the event stream (chapter 7) or we describe the full event model by ourselves.

We will just need to define protocol behavior (i.e. HTTP response codes and corresponding semantics) and have to decide if we want to adopt a textual event stream format or go with a serialized stream of JSON objects.
Since we have structured events in the TD, i.e. we have a JSON schema for event data,  the obvious alternative is to select an appropriate media type, e.g. JSON text sequences: https://datatracker.ietf.org/doc/html/rfc7464

Reply:
@mlagally wrote:
> Chaper 5 of this specification defines a **normative processing model** on the user agent, which assumes a JavaScript based browser implementation and mandates a specific implementation methodology e.g. describing task queues etc.
> which is appropriate for a browser.
> However these requirements put unnecessary requirements on non-browser consumers and should be excluded. We need however to incorporate the protocol handling part from that chapter, but these are only a few paragraphs.

Yes, that is what I meant when I said that the wire protocol is tightly coupled with the scripting API. In my experience of [implementing](https://github.com/WebThingsIO/gateway/pull/2863) the specification that wasn't really a problem, but it does make it tricky to cite as a reference.

> Chapter 7 defines 4 field names of the event stream and their interpretation
> This is a plain text serialisation of event types and event data where we need to define a mapping of the data schema, which is defined by the event affordance of the TD spec (section 5.3.1.5) to this serialised text stream.
> Event types are defined in SSE, however we need to defin a mapping as well.
> It would be appropriate to consider

The specification text I have written already does this:

>> Whenever an event of the specified type occurs while the Web Thing has an open connection with a Consumer, the Web Thing MUST send event data to the Consumer using the event stream format in the Server-Sent Events [EVENTSOURCE] specification. For each message sent, the Web Thing MUST set the event field to the name of the EventAffordance and populate the data field with event data, if any. The event data MUST follow the data schema specified in the EventAffordance and MUST be serialized in JSON. The id field SHOULD be set to a unique identifier for the event, for use when re-establishing a dropped connection (see below).

> The retry field leaves some room for interpretation (which unit?)

The unit is defined as milliseconds https://www.w3.org/TR/eventsource/#concept-event-stream-reconnection-time

> also the permitted values of the id field needs clarification.

The id can be any string, encoded as UTF-8. https://www.w3.org/TR/eventsource/#concept-event-stream-last-event-id In the implementations I've seen it's quite common to use a timestamp (number of milliseconds past the epoch) as the id, which is also what I used when I implemented it.

> Chapter 9 assumes a push proxy - we need to discuss details further.

I think this is just implementation advice, and shouldn't pose any problems. We should assume that all parts of the HTTP protocol binding could be proxied. In the case of events it is important for clients to send a `Cache-Control: no-cache` header to prevent proxies from caching responses, as defined in the specification.

> Chapter 10 is probably not required by the WoT profile.

Again, this is really just implementation advice for web browsers (`Window` and `Document` objects probably won't exist in other clients).

> We will just need to define protocol behavior (i.e. HTTP response codes and corresponding semantics)

I think in general we should leave this down to the Server-Sent Events specification to define, but is there a specific response code which you think needs defining because it's specific to our use case?

> and have to decide if we want to adopt a textual event stream format or go with a serialized stream of JSON objects.
> Since we have structured events in the TD, i.e. we have a JSON schema for event data, the obvious alternative is to select an appropriate media type, e.g. JSON text sequences: https://datatracker.ietf.org/doc/html/rfc7464

This is unnecessary. Each event payload consists of a single JSON text, which in the proposed specification text I quoted above is directly mapped onto the data field of the event stream format. It's very common to use JSON text as a data payload format for Server-Sent Events. If you tried to send a sequence of events as JSON text sequences rater than following the `text/event-stream` format, then existing clients (like web browsers) wouldn't understand them.

Now that we have established that trying to cherry-pick parts of the specification doesn't work very well because they all reference each other, I think we could just remove the note and reference the specification as a whole. I agree it's not perfect because the specification is clearly written with browser vendors in mind, but there are plenty of examples of non-JavaScript implementations where developers were able to successfully implement the wire protocol without the scripting API.